components/cloud/src/kms.rs

```diff
@@ -0,0 +1,155 @@
+use crate::error::{empty_key_contents, Result};
+use kvproto::encryptionpb::MasterKeyKms;
+use tokio::runtime::Runtime;
+
+#[derive(Debug, Clone)]
+pub struct Location {
+    pub region: String,
+    pub endpoint: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct Config {
+    pub key_id: KeyId,
+    pub location: Location,
+    pub vendor: String,
+}
+
+impl Config {
+    pub fn from_proto(mk: MasterKeyKms) -> Result<Self> {
+        Ok(Config {
+            key_id: KeyId::new(mk.key_id)?,
+            location: Location {
+                region: mk.region,
+                endpoint: mk.endpoint,
+            },
+            vendor: mk.vendor,
+        })
+    }
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct KeyId(String);
+
+// KeyID is a newtype to mark a String as an ID of a key
+// This ID exists in a foreign system such as AWS
+// The key id must be non-empty
+impl KeyId {
+    pub fn new(id: String) -> Result<KeyId> {
+        if id.is_empty() {
+            Err(empty_key_contents("KMS key id can not be empty"))
+        } else {
+            Ok(KeyId(id))
+        }
+    }
+}
+
+impl std::ops::Deref for KeyId {
+    type Target = String;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+// EncryptedKey is a newtype used to mark data as an encrypted key
+// It requires the vec to be non-empty
+#[derive(PartialEq, Clone, Debug)]
+pub struct EncryptedKey(Vec<u8>);
```

Comment:
Can we remove redundant types like `EncryptedKey` and `DataKeyPair` from encryption crate? They are not being used elsewhere.

Reply:
I am not sure what you mean: they are used in the encryption crate.

Reply:
I think you are pointing to something where the split between the two crates is not quite right. I will make some improvements here and see what can be made of it.

Reply:
I was able to stop passing the runtime around and cleanup the interface. I then followed up with the above comment about completely separating the two crates.

Reply:
anything progressed?

Reply:
These small structs are duplicated because neither crate depends on each other as per the other conversation. the encryption_export crate binds the two crates together. Duplication in general is a problem when using non type-safe languages (we are not) or when implementations that should be the same drift apart. However, our entire goal of the cloud library is to provide an interface that lets the cloud library make changes independently of TiKV. So we do not want the implementations to be the same. There is already some difference in EncryptedKey that cannot be resolved by importing it from cloud to encryption. We very much want to allow these definitions to drift apart as needed. The cloud just needs to maintain the interface: `EncryptedKey::new` (takes a vec) and the two DataKeyPair fields support `.to_vec`.